File: legacy/src/main/rs/emulate/legacy/config/graphic/DefaultGraphicDefinition.java
package rs.emulate.legacy.config.graphic;

import rs.emulate.legacy.config.ConfigConstants;
import rs.emulate.legacy.config.ConfigUtils;
import rs.emulate.legacy.config.DefaultConfigDefinition;
import rs.emulate.legacy.config.SerializableProperty;

import java.util.HashMap;
import java.util.Map;

import static rs.emulate.legacy.config.Properties.unsignedByte;
import static rs.emulate.legacy.config.Properties.unsignedShort;
import static rs.emulate.legacy.config.graphic.GraphicDefinition.COLOUR_COUNT;
import static rs.emulate.legacy.config.graphic.GraphicProperty.*;

/**
 * A default {@link GraphicDefinition} used as a base for an actual definition.
 *
 * @param <T> The type of GraphicDefinition this DefaultGraphicDefinition is for.
 * @author Major
 */
public class DefaultGraphicDefinition<T extends GraphicDefinition> extends DefaultConfigDefinition<T> {

	@Override
	protected Map<Integer, SerializableProperty<?>> init() {
		Map<Integer, SerializableProperty<?>> defaults = new HashMap<>(27);

		defaults.put(1, unsignedShort(MODEL, 0));
		defaults.put(2, unsignedShort(ANIMATION, -1));
		defaults.put(4, unsignedShort(BREADTH_SCALE, ConfigConstants.DEFAULT_SCALE));
		defaults.put(5, unsignedShort(DEPTH_SCALE, ConfigConstants.DEFAULT_SCALE));
		defaults.put(6, unsignedShort(ROTATION, 0));
		defaults.put(7, unsignedByte(BRIGHTNESS, 0));
		defaults.put(8, unsignedByte(SHADOW, 0));

		for (int slot = 1; slot <= COLOUR_COUNT; slot++) {
			defaults.put(slot + 40, unsignedShort(ConfigUtils.getOriginalColourPropertyName(slot), 0));
			defaults.put(slot + 50, unsignedShort(ConfigUtils.getReplacementColourPropertyName(slot), 0));
		}

		return defaults;
	}

}

File: legacy/src/main/kotlin/rs/emulate/legacy/config/graphic/DefaultGraphicDefinition.kt
package rs.emulate.legacy.config.graphic

import rs.emulate.legacy.config.ConfigConstants
import rs.emulate.legacy.config.ConfigUtils
import rs.emulate.legacy.config.DefaultConfigDefinition
import rs.emulate.legacy.config.Properties.unsignedByte
import rs.emulate.legacy.config.Properties.unsignedShort
import rs.emulate.legacy.config.SerializableProperty
import rs.emulate.legacy.config.graphic.GraphicDefinition.Companion.COLOUR_COUNT
import rs.emulate.legacy.config.graphic.GraphicProperty.*
import java.util.HashMap

/**
 * A default [GraphicDefinition] used as a base for an actual definition.
 */
class DefaultGraphicDefinition : DefaultConfigDefinition<GraphicDefinition>() {

    override fun init(): Map<Int, SerializableProperty<*>> {
        val defaults = HashMap<Int, SerializableProperty<*>>(27)

        defaults[1] = unsignedShort(Model, 0)
        defaults[2] = unsignedShort(Animation, -1)
        defaults[4] = unsignedShort(BreadthScale, ConfigConstants.DEFAULT_SCALE)
        defaults[5] = unsignedShort(DepthScale, ConfigConstants.DEFAULT_SCALE)
        defaults[6] = unsignedShort(Rotation, 0)
        defaults[7] = unsignedByte(Brightness, 0)
        defaults[8] = unsignedByte(Shadow, 0)

        for (slot in 0 until COLOUR_COUNT) {
            defaults[slot + 40] = unsignedShort(ConfigUtils.getOriginalColourPropertyName(slot, 40), 0)
            defaults[slot + 50] = unsignedShort(ConfigUtils.getReplacementColourPropertyName(slot, 50), 0)
        }

        return defaults
    }

}
